backend/de.metas.swat/de.metas.swat.base/src/main/java/org/adempiere/invoice/process/C_Invoice_ReissueInvoice.java

```diff
@@ -0,0 +1,159 @@
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.compiere.model.I_C_DocType;
+import org.compiere.model.I_C_Invoice;
+
+import java.util.Optional;
+
+import static org.compiere.model.X_C_DocType.DOCBASETYPE_ARInvoice;
+
+/**
+ * TODO
+ */
+public class C_Invoice_ReissueInvoice extends JavaProcess implements IProcessPrecondition
+{
+	private final IInvoiceDAO invoiceDAO = Services.get(IInvoiceDAO.class);
+	private final IInvoiceBL invoiceBL = Services.get(IInvoiceBL.class);
+	private final IDocTypeDAO docTypeDAO = Services.get(IDocTypeDAO.class);
+	private final ICountryDAO countryDAO = Services.get(ICountryDAO.class);
+	private final INotificationBL notificationBL = Services.get(INotificationBL.class);
+	private final IUserBL userBL = Services.get(IUserBL.class);
+
+	private static final DocBaseAndSubType SALES_INVOICE = DocBaseAndSubType.of(DOCBASETYPE_ARInvoice);
+	private static final AdMessageKey MSG_Event_DocumentGenerated = AdMessageKey.of("Event_DocumentGenerated");
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(final @NonNull IProcessPreconditionsContext context)
+	{
+		if (!context.isSingleSelection())
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNotSingleSelection().toInternal();
+		}
+
+		final InvoiceId invoiceId = InvoiceId.ofRepoId(context.getSingleSelectedRecordId());
+		final I_C_Invoice invoiceRecord = invoiceDAO.getByIdInTrx(invoiceId);
+		final DocTypeId docTypeId = DocTypeId.ofRepoId(invoiceRecord.getC_DocType_ID());
+		final DocBaseAndSubType docBaseAndSubType = docTypeDAO.getDocBaseAndSubTypeById(docTypeId);
+
+		if(!docBaseAndSubType.equals(SALES_INVOICE))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not an basic Sales Invoice");
+		}
+
+		final DocStatus docStatus = DocStatus.ofCode(invoiceRecord.getDocStatus());
+		if(!docStatus.isCompletedOrClosed())
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("Not completed or closed");
+		}
+
+		if(countryDAO.isEnforceCorrectionInvoice(invoiceBL.getBillToCountryId(invoiceId)))
+		{
+			return ProcessPreconditionsResolution.rejectWithInternalReason("isEnforceCorrectionInvoice");
+		}
+
+		return ProcessPreconditionsResolution.accept();
+	}
+
+	@Override
+	protected String doIt() throws Exception
+	{
+		final de.metas.adempiere.model.I_C_Invoice invoice = InterfaceWrapperHelper.create(getCtx(), getRecord_ID(), de.metas.adempiere.model.I_C_Invoice.class, get_TrxName());
+
+		final InvoiceCreditContext creditCtx = InvoiceCreditContext.builder()
+				.docTypeId(null)
+				.completeAndAllocate(true)
+				.referenceOriginalOrder(true)
+				.referenceInvoice(true)
+				.creditedInvoiceReinvoicable(false)
+				.fixedInvoice(true)
+				.build();
+
+		final I_C_Invoice creditMemo = invoiceBL.creditInvoice(invoice, creditCtx);
+		createUserNotification(creditMemo);
+
+		final de.metas.adempiere.model.I_C_Invoice salesInvoice = InterfaceWrapperHelper.create(
```

Comment:
why do u have to use InterfaceWrapperHelper.create to "cast" the returned invoice?

pls consider changing the EntityType of C_Invoice.Ref_Invoice_ID to `D`ictionary and generate the C_Invoice model so you will get the setRef_Invoice_ID in the main model 

(if that's the reason)

Reply:
honestly i copied it from `org.adempiere.invoice.process.CreateCreditMemoFromInvoice` and i didn't feel comfortable to completely refactor `de.metas.invoice.service.IInvoiceBL#creditInvoice()` (AbstractBL)

Reply:
ok